draft-irtf-cfrg-bbs-signatures.md

```diff
@@ -779,8 +787,158 @@ Procedure:
 11. return (scalar_1, ..., scalar_count)
 ```
 
+## Domain Calculation
+
+This operation calculates the domain value, a scalar encoding public parameter information. Optionally, the domain can also include an additional octet string, called the header, meant to encode any additional information the signer may want to "bind" to the signature (like an expiration date, the targeted audience identifier etc.). The domain value will be signed using a specific generator point (`Q_2`) (see [Sign](#sign)) to protect the integrity of those public parameters. The header, if existing, must be known by all parties (i.e., the signer, the prover and verifier), for both the signature and proof to be validated correctly. This means, that the header can be used to enforce certain values (chosen by the signer) to always be disclosed by the prover.
+
+This operation makes use of the `serialize` function, defined in [Section 4.6.1](#serialize).
+
+```
+domain = get_domain(Q_1, Q_2, H_Points, PK, header)
+
+Inputs:
+
+- PK (REQUIRED), octet string, representing the public key of the
+                 Signer of the form outputted by the SkToPk operation.
+- (Q_1, Q_2) (REQUIRED), points of G1 (the first 2 points returned from
+                         create_generators).
+- H_Points (REQUIRED), array of points of G1.
+- header (OPTIONAL), octet string. If not supplied, it must default to
+                     the empty octet string ("").
+
+Parameters:
+
+- ciphersuite_id, ASCII string. The unique ID of the ciphersuite.
+
+Outputs:
+
+- domain, a scalar value or INVALID.
+
+Procedure:
+
+1. L = length(H_Points)
+2. if length(header) > 2^64 - 1 or L > 2^64 - 1, return INVALID
+3. (H_1, ..., H_L) = H_Points
+4. dom_array = (Q_1, Q_2, L, H_1, ..., H_L, ciphersuite_id)
+5. dom_octs = serialize(dom_array)
+6. if dom_octs is INVALID, return INVALID
+7. dom_input = PK || dom_octs || I2OSP(length(header), 8) || header
+8. return hash_to_scalar(dom_input, 1)
+```
+
+**Note**: If the header is not supplied in `get_domain`, it defaults to the empty octet string (""). This means that in the concatenation step of the above procedure (step 7), 8 bytes representing a length of 0 (i.e., `0x0000000000000000`), will still need to be appended at the end, even though the a header value is not provided.
```

Comment:
```suggestion
**Note**: If the header is not supplied in `get_domain`, it defaults to the empty octet string ("").
```

I think it's clear in the algorithm that the length is always included

Reply:
That's true, though IMO it would be an easy mistake to make.

its clear only if you notice the `If not supplied, it must default to the empty octet string ("")` part. If someone misses that and only notice the capitalized `(OPTIONAL)` could maybe make this mistake??

Reply:
maybe we could update the text to say `If not supplied, it must default to the empty octet string ("") **that has 0 length**`